examples/batched-solver/batched-solver.cpp

```diff
@@ -0,0 +1,378 @@
+        }
+    }
+    if (print_time) {
+        std::cout << apply_time / num_reps << std::endl;
+    } else {
+        std::cout << "Solver type: "
+                  << "batch::bicgstab"
+                  << "\nMatrix size: " << A->get_common_size()
+                  << "\nNum batch entries: " << A->get_num_batch_items()
+                  << "\nEntire solve took: " << apply_time / num_reps
+                  << " seconds." << std::endl;
+    }
+
+    // Ginkgo objects are cleaned up automatically; but the "application" still
+    //  needs to clean up its data in this case.
+    appl_clean_up(appl_sys, exec);
+    return 0;
+}
+
+ApplSysData appl_generate_system(const int nrows, const size_type nsystems,
+                                 std::shared_ptr<gko::Executor> exec)
+{
+    const int nnz = nrows * 3 - 2;
+    std::ranlux48 rgen(15);
+    std::normal_distribution<real_type> distb(0.5, 0.1);
+    std::vector<real_type> spacings(nsystems * nrows);
+    std::generate(spacings.begin(), spacings.end(),
+                  [&]() { return distb(rgen); });
+
+    std::vector<value_type> allvalues(nnz * nsystems);
+    for (size_type isys = 0; isys < nsystems; isys++) {
+        allvalues[isys * nnz] = 2.0 / spacings[isys * nrows];
+        allvalues[isys * nnz + 1] = -1.0;
+        for (int irow = 0; irow < nrows - 2; irow++) {
+            allvalues[isys * nnz + 2 + irow * 3] = -1.0;
+            allvalues[isys * nnz + 2 + irow * 3 + 1] =
+                2.0 / spacings[isys * nrows + irow + 1];
+            allvalues[isys * nnz + 2 + irow * 3 + 2] = -1.0;
+        }
+        allvalues[isys * nnz + 2 + (nrows - 2) * 3] = -1.0;
+        allvalues[isys * nnz + 2 + (nrows - 2) * 3 + 1] =
+            2.0 / spacings[(isys + 1) * nrows - 1];
+        assert(isys * nnz + 2 + (nrows - 2) * 3 + 2 == (isys + 1) * nnz);
+    }
+
+    std::vector<index_type> rowptrs(nrows + 1);
+    rowptrs[0] = 0;
+    rowptrs[1] = 2;
+    for (int i = 2; i < nrows; i++) {
+        rowptrs[i] = rowptrs[i - 1] + 3;
+    }
+    rowptrs[nrows] = rowptrs[nrows - 1] + 2;
+    assert(rowptrs[nrows] == nnz);
+
+    std::vector<index_type> colidxs(nnz);
+    colidxs[0] = 0;
+    colidxs[1] = 1;
+    const int nnz_per_row = 3;
+    for (int irow = 1; irow < nrows - 1; irow++) {
+        colidxs[2 + (irow - 1) * nnz_per_row] = irow - 1;
+        colidxs[2 + (irow - 1) * nnz_per_row + 1] = irow;
+        colidxs[2 + (irow - 1) * nnz_per_row + 2] = irow + 1;
+    }
+    colidxs[2 + (nrows - 2) * nnz_per_row] = nrows - 2;
+    colidxs[2 + (nrows - 2) * nnz_per_row + 1] = nrows - 1;
+    assert(2 + (nrows - 2) * nnz_per_row + 1 == nnz - 1);
+
+    std::vector<value_type> allb(nrows * nsystems);
+    for (size_type isys = 0; isys < nsystems; isys++) {
+        const value_type bval = distb(rgen);
+        std::fill(allb.begin() + isys * nrows,
+                  allb.begin() + (isys + 1) * nrows, bval);
+    }
+
+    index_type* const row_ptrs = exec->alloc<index_type>(nrows + 1);
```

Comment:
I think we will operate on the application memory not our own (as they can give us the non-const memory with current const limitation)

Reply:
Yes, this function is meant to emulate the application.